proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -67,101 +68,597 @@ store values that may only be determined at runtime.
-      32, 32, MATRIX_LAYOUT_INFERENCING_OPTIMAL>
-      matrix1 = {model, 2048, 0};
+  MatrixRef<DATA_TYPE_E4M3, 32, 32, MATRIX_LAYOUT_MUL_OPTIMAL> Matrix1 = {
+      Model, 2048, 0};
 
-  VectorRef<DATA_TYPE_FLOAT16> biasVector1 = {model, 3072};
+  VectorRef<DATA_TYPE_FLOAT16> BiasVector1 = {Model, 3072};
 
-  MatrixRef<
-      DATA_TYPE_E4M3,
-      3, 32, MATRIX_LAYOUT_INFERENCING_OPTIMAL>
-      matrix2 = {model, 4096, 0};
+  MatrixRef<DATA_TYPE_E4M3, 3, 32, MATRIX_LAYOUT_MUL_OPTIMAL> Matrix2 = {
+      Model, 4096, 0};
 
-  VectorRef<DATA_TYPE_FLOAT16> biasVector2 = {model, 5120};
+  VectorRef<DATA_TYPE_FLOAT16> BiasVector2 = {Model, 5120};
 
-  vector<half, 32> layer0 = MulAdd<half>(
-      matrix0,
-      InterpretedVector<DATA_TYPE_E4M3>(inputVector),
-      biasVector0);
-  layer0 = max(layer0, 0);
+  vector<half, 32> Layer0 = MulAdd<half>(
+      Matrix0, MakeInterpretedVector<DATA_TYPE_E4M3>(InputVector), BiasVector0);
+  Layer0 = max(Layer0, 0);
 
-  vector<half, 32> layer1 = MulAdd<half>(
-      matrix1,
-      InterpretedVector<DATA_TYPE_E4M3>(layer0),
-      biasVector1);
-  layer1 = max(layer1, 0);
+  vector<half, 32> Layer1 = MulAdd<half>(
+      Matrix1, MakeInterpretedVector<DATA_TYPE_E4M3>(Layer0), BiasVector1);
+  Layer1 = max(Layer1, 0);
 
-  vector<float, 3> output = MulAdd<float>(
-      matrix2,
-      InterpretedVector<DATA_TYPE_E4M3>(layer1),
-      biasVector2);
-  output = exp(output);
+  vector<float, 3> Output = MulAdd<float>(
+      Matrix2, MakeInterpretedVector<DATA_TYPE_E4M3>(Layer1), BiasVector2);
+  Output = exp(Output);
 
-  return output;
+  return Output;
 }
 ```
 
 ## Detailed design
 
-TBD
+The API is implemented using the C++03 style templates supported in HLSL 2021.
+Therefore HLSL 2021 is required to use this API.
+
+Developers are expected to rely on template argument deduction for calls to the
+various functions defined. In this document a higher-level view of how
+developers should view each function is provided, along with a proposed
+implementation.
 
 ### Builtins
 
 TBD
 
+### enum DataType
+
+The `dx::linalg::DataType` enum defines the various data types that can be
+applied to matrices and vectors. The numeric values of these enum entries are
+picked to match those used in the underlying DXIL.
+
+```c++
+namespace dx {
+namespace linalg {
+
+enum DataType {
+  DATA_TYPE_SINT16 = 2,           // ComponentType::I16
+  DATA_TYPE_UINT16 = 3,           // ComponentType::U16
+  DATA_TYPE_SINT32 = 4,           // ComponentType::I32
+  DATA_TYPE_UINT32 = 5,           // ComponentType::U32
+  DATA_TYPE_FLOAT16 = 8,          // ComponentType::F16
+  DATA_TYPE_FLOAT32 = 9,          // ComponentType::F32
+  DATA_TYPE_SINT8_T4_PACKED = 17, // ComponentType::PackedS8x32
+  DATA_TYPE_UINT8_T4_PACKED = 18, // ComponentType::PackedU8x32
+  DATA_TYPE_UINT8 = 19,           // ComponentType::U8
+  DATA_TYPE_SINT8 = 20,           // ComponentType::I8
+  DATA_TYPE_E4M3 = 21,            // ComponentType::F8_E4M3
```

Comment:
Should we add 'F8' to have DATA_TYPE_F8_E4M3 and DATA_TYPE_F8_E5M2 ?

Reply:
Renamed to DATA_TYPE_FLOAT8_E4M3 and DATA_TYPE_FLOAT8_E5M2, to be consistent with the FLOAT32 and FLOAT16 ones.